### PR TITLE
Add methods to insert and display interviews

### PR DESCRIPTION
**Pull request by YANLING-H**
### Task completed
- add use case and view model methods to insert interviews to the database and display interviews in the dashboard screen.

### Tests
https://github.com/Audienix/InterPrep/assets/67870437/4c27afca-dff0-42c4-b6f9-b22a748123b4

### CODE REVIEW

File: app/src/main/kotlin/com/twain/interprep/data/model/Interview.kt [hunk 1]
```diff
@@ -7,16 +7,16 @@ import java.util.*
...
-    val interviewType: String?,
-    val role: String?,
-    val roundNum: Int?,
-    val jobPostLink: String?,
-    val companyLink: String?,
-    val interviewer: String?,
-    val interviewStatus: InterviewStatus
+    val interviewType: String? = "",
+    val role: String? = "",
+    val roundNum: Int? = null,
```
Comment (Audienix) [inline app/src/main/kotlin/com/twain/interprep/data/model/Interview.kt]: Should not default value be 1 here? Or if you are keeping null, then we should not show null in UI in dashboard interview cards, as I see from the PR video.

File: app/src/main/kotlin/com/twain/interprep/domain/usecase/interview/GetInterviewsUseCase.kt [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+                    dashBoardInterviews.upcomingInterviews.add(interview)
+                } else {
+                    dashBoardInterviews.comingNextInterviews.add(interview)
+                }
+            }
+            emit(dashBoardInterviews)
+        }
+
+
+    private fun isSameWeekAsCurrentDate(date: Date): Boolean {
```
Comment (Audienix) [inline app/src/main/kotlin/com/twain/interprep/domain/usecase/interview/GetInterviewsUseCase.kt]: nit: Naming suggestion
```suggestion
    private fun isInterviewDateInSameWeek(date: Date): Boolean {
```

File: app/src/main/kotlin/com/twain/interprep/domain/usecase/interview/GetInterviewsUseCase.kt [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+        val current = Calendar.getInstance()
+        val comparedCalendar = Calendar.getInstance()
```
Comment (Audienix) [inline app/src/main/kotlin/com/twain/interprep/domain/usecase/interview/GetInterviewsUseCase.kt]: nit: Naming Suggestion
```suggestion
        val currentDateTime = Calendar.getInstance()
        val interviewDateTime = Calendar.getInstance()
```

File: app/src/main/kotlin/com/twain/interprep/domain/usecase/interview/GetInterviewsUseCase.kt [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+        current.time = Date()
+        comparedCalendar.time = date
+
+        val currentWeek = current.get(Calendar.WEEK_OF_YEAR)
+        val comparedWeek = comparedCalendar.get(Calendar.WEEK_OF_YEAR)
+
+        val currentYear = current.get(Calendar.YEAR)
+        val comparedYear = comparedCalendar.get(Calendar.YEAR)
+
+        return currentWeek == comparedWeek && currentYear == comparedYear
```
Comment (Audienix) [inline app/src/main/kotlin/com/twain/interprep/domain/usecase/interview/GetInterviewsUseCase.kt:51]: Want to confirm this logic - interview date's calendar week is equal to current calendar week and also both date falls in same year then we are considering interview is Upcoming Interview ?
If that is so then 
1. how about a case when an interview falls on 2nd January on Thursday and I am seeing the app on 30th December on Monday ? Will I see the 2nd Jan interview as upcoming ? We should see that as upcoming.

2. Consider adding device locale while setting up the calendar. Calendar.getInstance(Locale.getDefault())

File: app/src/main/kotlin/com/twain/interprep/presentation/ui/modules/interview/InterviewViewModel.kt [hunk 5]
```diff
@@ -1,36 +1,60 @@
...
 import com.twain.interprep.presentation.ui.modules.common.BaseViewModel
 import com.twain.interprep.helper.CoroutineContextDispatcher
 import dagger.hilt.android.lifecycle.HiltViewModel
 import kotlinx.coroutines.CoroutineExceptionHandler
+import kotlinx.coroutines.flow.collect
+import java.util.Date
 import javax.inject.Inject
 
 @HiltViewModel
 class InterviewViewModel @Inject constructor(
```
Comment (Audienix) [inline app/src/main/kotlin/com/twain/interprep/presentation/ui/modules/interview/InterviewViewModel.kt:19]: If you wanna keep `InterviewViewModel` separate from `DashboardViewModel` then consider moving `AddInterviewScreen` under `interview` package.

File: app/src/main/kotlin/com/twain/interprep/util/InterviewsData.kt [hunk 6]
```diff
@@ -0,0 +1,38 @@
+package com.twain.interprep.util
+
+import com.twain.interprep.data.model.Interview
+import java.text.SimpleDateFormat
+import java.time.LocalDate
+import java.util.Date
+
+class InterviewsData {
```
Comment (Audienix) [inline app/src/main/kotlin/com/twain/interprep/util/InterviewsData.kt]: I could not understand the significance of this class.

File: app/src/main/kotlin/com/twain/interprep/presentation/ui/modules/interview/InterviewViewModel.kt [hunk 7]
```diff
@@ -1,36 +1,60 @@
 package com.twain.interprep.presentation.ui.modules.interview
 
+import androidx.compose.runtime.MutableState
+import androidx.compose.runtime.getValue
+import androidx.compose.runtime.mutableStateOf
+import androidx.compose.runtime.setValue
 import com.twain.interprep.data.model.Interview
+import com.twain.interprep.data.model.ViewResult
 import com.twain.interprep.domain.repository.InterviewRepository
```
Comment (Audienix) [inline app/src/main/kotlin/com/twain/interprep/presentation/ui/modules/interview/InterviewViewModel.kt]: unused imports

File: app/src/main/kotlin/com/twain/interprep/presentation/ui/modules/dashboard/DashboardScreen.kt [hunk 8]
```diff
@@ -28,14 +32,21 @@ import com.twain.interprep.presentation.ui.components.PastInterviewCard
...
 @Composable
 fun DashboardScreen(
     navController: NavHostController,
-    viewModel: DashboardViewModel = hiltViewModel(),
+    viewModel: InterviewViewModel = hiltViewModel(),
     quotesViewModel: QuotesViewModel = hiltViewModel()
 ) {
+    // TODO ask Arighna
     quotesViewModel.insertQuotes(QuoteData.quotes)
+    LaunchedEffect(Unit){
```
Comment (Audienix) [inline app/src/main/kotlin/com/twain/interprep/presentation/ui/modules/dashboard/DashboardScreen.kt]: Do we require LaunchedEffect ?

LaunchedEffect: run suspend functions in the scope of a composable

To call suspend functions safely from inside a composable, use the [LaunchedEffect](https://developer.android.com/reference/kotlin/androidx/compose/runtime/package-summary#LaunchedEffect(kotlin.Any,kotlin.coroutines.SuspendFunction1)) composable. When LaunchedEffect enters the Composition, it launches a coroutine with the block of code passed as a parameter. The coroutine will be cancelled if LaunchedEffect leaves the composition.

File: app/src/main/kotlin/com/twain/interprep/presentation/ui/modules/dashboard/DashboardScreen.kt [hunk 9]
```diff
@@ -54,58 +65,67 @@ fun DashboardScreen(
...
-                        LazyRow(
-                            modifier = Modifier.padding(bottom = dimensionResource(id = R.dimen.dimension_8dp)),
-                            verticalAlignment = Alignment.CenterVertically
-                        ) {
-                            items(5) { UpcomingInterviewCard() }
+                if (viewModel.interviews is ViewResult.Loaded){
+                    val interviews = viewModel.interviews as ViewResult.Loaded
+                    item {
+                        Column {
+                            IPHeader(
```
Comment (Audienix) [inline app/src/main/kotlin/com/twain/interprep/presentation/ui/modules/dashboard/DashboardScreen.kt]: Any section header should not be visible if there is no data for that section.

File: app/src/main/kotlin/com/twain/interprep/presentation/ui/modules/dashboard/AddInterviewScreen.kt [hunk 10]
```diff
@@ -43,7 +48,16 @@ fun AddInterviewScreen(
         topBar = {
             IPAppBar(stringResource(id = R.string.appbar_title_add_interview)) {
                 IconButton(onClick = {
+
+                    viewModel.insertInterview(
+                        Interview(
```
Comment (Audienix) [inline app/src/main/kotlin/com/twain/interprep/presentation/ui/modules/dashboard/AddInterviewScreen.kt]: We have to show existing AlertDialog component if mandatory fields are empty.

**Review by Audienix — COMMENTED** [review]
Left few comments. Will approve after those are addressed or clarified.

File: app/src/main/kotlin/com/twain/interprep/presentation/ui/modules/dashboard/DashboardScreen.kt (hunk 8 above)
Comment (Arighna88) [reply app/src/main/kotlin/com/twain/interprep/presentation/ui/modules/dashboard/DashboardScreen.kt]: Although it does not harm to use LaunchedEffect for one time job. So ignore my previous comment.

File: app/src/main/kotlin/com/twain/interprep/presentation/ui/modules/interview/InterviewViewModel.kt (hunk 5 above)
Comment (sev3rus) [reply app/src/main/kotlin/com/twain/interprep/presentation/ui/modules/interview/InterviewViewModel.kt:19]: We actually consider interview viewmodel and dashboard viewmodel to be the same viewmodel, we can remove one of them

File: app/src/main/kotlin/com/twain/interprep/util/InterviewsData.kt (hunk 6 above)
Comment (sev3rus) [reply app/src/main/kotlin/com/twain/interprep/util/InterviewsData.kt]: This class is for mock data only, I believe I will delete this in my later PR

File: app/src/main/kotlin/com/twain/interprep/presentation/ui/modules/interview/InterviewViewModel.kt (hunk 7 above)
Comment (sev3rus) [reply app/src/main/kotlin/com/twain/interprep/presentation/ui/modules/interview/InterviewViewModel.kt]: Uhhh this is used for later PR

File: app/src/main/kotlin/com/twain/interprep/data/model/Interview.kt (hunk 1 above)
Comment (YANLING-H) [reply app/src/main/kotlin/com/twain/interprep/data/model/Interview.kt]: To fix this problem, I made a small change in the` interviewCard.kt` to map undefined round to "unknown" round.

File: app/src/main/kotlin/com/twain/interprep/domain/usecase/interview/GetInterviewsUseCase.kt (hunk 4 above)
Comment (YANLING-H) [reply app/src/main/kotlin/com/twain/interprep/domain/usecase/interview/GetInterviewsUseCase.kt:51]: I believe the Calendar object handled the first issue as tested below:

File: app/src/main/kotlin/com/twain/interprep/presentation/ui/modules/dashboard/AddInterviewScreen.kt (hunk 10 above)
Comment (YANLING-H) [reply app/src/main/kotlin/com/twain/interprep/presentation/ui/modules/dashboard/AddInterviewScreen.kt]: I will create another pr to link the alertDialog.

File: app/src/main/kotlin/com/twain/interprep/domain/usecase/interview/GetInterviewsUseCase.kt (hunk 4 above)
Comment (Audienix) [reply app/src/main/kotlin/com/twain/interprep/domain/usecase/interview/GetInterviewsUseCase.kt:51]: @YANLING-H Your example shows both dates in same year. My 1st case is when both dates are in different year but in same week.

**Comment by Audienix** [comment]
@YANLING-H  @sev3rus I have fixed almost all the issues and removed lot of unnecessary code. 
1> AlertDialog for not filling up mandatory details - Can be handled in next PR.
2> I am approving this but plz fix the below issue in next PR.
 Issue - I added one interview at 12:53 AM 8th October. But it went to Past Interview. It is very easily reproducible issue. 

Screenshot 2023-06-08 at 12 53 07 AM

**Review by Audienix — APPROVED** [review]

**Pull request merged**